middleware/grpc_instrumentation.go

```diff
@@ -174,17 +165,26 @@ func (s *instrumentedClientStream) CloseSend() error {
 }
 
 // errorCode converts an error into an error code string.
-func errorCode(err error) string {
+func errorCode(err error, successCode string, maskHTTPStatuses bool) string {
 	if err == nil {
-		return "2xx"
+		return successCode
 	}
 
-	if errResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-		statusFamily := int(errResp.Code / 100)
-		return strconv.Itoa(statusFamily) + "xx"
-	} else if grpcutil.IsCanceled(err) {
+	if grpcutil.IsCanceled(err) {
 		return "cancel"
-	} else {
-		return "error"
 	}
+
+	if errResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
+		if maskHTTPStatuses {
+			statusFamily := int(errResp.Code / 100)
+			return strconv.Itoa(statusFamily) + "xx"
+		}
+		return strconv.Itoa(int(errResp.Code))
+	}
+
+	if stat, ok := grpcutil.ErrorToStatus(err); ok {
+		return stat.Code().String()
+	}
```

Comment:
This implementation seems inefficient. First we call `httpgrpc.HTTPResponseFromError` which calls `grpcutil.ErrorToStatus` and checks for "details". If it finds that error is not "HTTPResponse", result is thrown away, only to call `grpcutil.ErrorToStatus` again here.

I'd suggest to intruduce new `httpgrpc.HTTPResponseAndStatusFromError` that returns `*HTTPResponse` (possibly nil) and also `*status.Status` (possibly nil) in one go. Then we don't need another call to `grpcutil.ErrorToStatus(err)` here.

Reply:
WDYT about this:
```
// errorCode converts an error into an error code string.
func errorCode(err error, successCode string, maskHTTPStatuses bool) string {
	if err == nil {
		return successCode
	}

	if grpcutil.IsCanceled(err) {
		return "cancel"
	}

	if stat, ok := grpcutil.ErrorToStatus(err); ok {
		statusFamily := int(stat.Code() / 100)
		if statusFamily > 0 && maskHTTPStatuses {
			return strconv.Itoa(statusFamily) + "xx"
		}
		return stat.Code().String()
	}

	return "error"
}
```
At the end, we don't care whether it is a `httpgrpc` or a pure `gRPC` error, we just care whether it can be converted to the `status` interface or not, and get the corresponding code. This way we reduce the number of calls to `FromError()`

Reply:
This seems like a hack, since we're not checking for HTTPResponse at all.

We also need to do this:

```golang
		if statusFamily > 0 {
			// codes bigger than 100 are expected to be HTTP status codes.
			if maskHTTPStatuses {
				return strconv.Itoa(statusFamily) + "xx"
			} 
			return strconv.Itoa(stat.Code())
		}
```

Your example shows why it makes sense to have `ErrorToStatusCode` only -- again, there's no need to do grpc -> gogo status conversion.

Reply:
I just realized that if we inline `grpcutil.IsCanceled` here, we get rid of yet-another conversion.

Reply:
> I just realized that if we inline grpcutil.IsCanceled here, we get rid of yet-another conversion.

I can do that.

Reply:
> Your example shows why it makes sense to have ErrorToStatusCode only -- again, there's no need to do grpc -> gogo status conversion.

I am afraid I dont' understand your remark about grpc -> gogo status conversion.
My concerns about `ErrorToStatusCode` are:
1) status codes of `HTTPResponse` are of type `int32`
2) both `gogo` and `grpc` statuses have codes of type `codes.Code`
3) we can return the string representation of them, but we need to distinguish between HTTP statuses and gRPC statuses, because the former need to be masked in some cases (e.g., `4xx`)
4) the conversion is needed if we receive, for example, a `DoNotLogError{Err: gogostatus.New(code.Internal, "error")}`. In this cases `gogostatus.FromError()` will not recognize the error